src/metadata_key.rs

```diff
@@ -0,0 +1,249 @@
+use bytes::Bytes;
+use http;
+use http::header::HeaderName;
+
+use std::borrow::Borrow;
+use std::error::Error;
+use std::fmt;
+use std::str::FromStr;
+
+/// Represents a custom metadata field name.
+///
+/// `MetadataKey` is used as the [`MetadataMap`] key.
+///
+/// [`HeaderMap`]: struct.HeaderMap.html
+#[derive(Clone, Eq, PartialEq, Hash)]
+pub struct MetadataKey {
+    pub(crate) inner: http::header::HeaderName,
+}
+
+/// A possible error when converting a `MetadataKey` from another type.
+#[derive(Debug)]
+pub struct InvalidMetadataKey {
+    _priv: (),
+}
+
+impl MetadataKey {
+    /// Converts a slice of bytes to a `MetadataKey`.
+    ///
+    /// This function normalizes the input.
+    pub fn from_bytes(src: &[u8]) -> Result<MetadataKey, InvalidMetadataKey> {
+        match HeaderName::from_bytes(src) {
+            Ok(name) => Ok(MetadataKey { inner: name }),
+            Err(_) => Err(InvalidMetadataKey::new())
+        }
+    }
+
+    /// Converts a slice of bytes to a `MetadataKey`.
+    ///
+    /// This function expects the input to only contain lowercase characters.
+    /// This is useful when decoding HTTP/2.0 headers. The HTTP/2.0
+    /// specification requires that all headers be represented in lower case.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tower_grpc::metadata::*;
+    ///
+    /// // Parsing a lower case metadata key
+    /// let hdr = MetadataKey::from_lowercase(b"content-length").unwrap();
+    /// assert_eq!("content-length", hdr);
+    ///
+    /// // Parsing a metadata key that contains uppercase characters
+    /// assert!(MetadataKey::from_lowercase(b"Content-Length").is_err());
+    /// ```
+    pub fn from_lowercase(src: &[u8]) -> Result<MetadataKey, InvalidMetadataKey> {
```

Comment:
I would delete this for now. It was added to better support HTTP/2.0 implementations. gRPC should not need it.

Reply:
Ok, will remove.